### ISSUE TITLE
Training T5 from scratch on English (no pre-training)

### ISSUE THREAD

**Issue by chrisrytting** [google-research/text-to-text-transfer-transformer#387]
I'm trying to compare sample efficiency of pre-trained T5 and from-scratch T5 on a custom task (for the sake of a baseline in a paper). In looking for examples of this, I can only find either (a) the general approach: fine-tuning the pre-trained models you guys have shared in order to perform tasks in English or (b) the new language approach: training a new tokenizer on some other language and doing pre-training there before then fine-tuning.

But what if I don't want pre-training at all? Is there a simple way to train from scratch? When I try to do this, I simply modify my fine-tuning script to invoke the `train` command (from `MtfModel`) on a new model directory called 3Bscratch (which contains, at this point, only the operative_config.gin file that I copied from the pretrained model directory, which I've been able to fine-tune on) instead of the `finetune` command (also from `MtfModel`), and I get the following error:

```
RuntimeError: Required bindings for `make_layer_stack` not provided in config: ['layers']
  In call to configurable 'make_bitransformer' (<function make_bitransformer at 0x7f1141bd2d90>)
```

**Comment by craffel** [comment]
See this section of the README: https://github.com/google-research/text-to-text-transfer-transformer#training

Based on your error, I think you have not provided a model configuration via gin.

**Comment by chrisrytting** [comment]
@craffel in the issue I mentioned that "I simply modify my fine-tuning script to invoke the train command (from MtfModel) on a new model directory called 3Bscratch (which contains, at this point, only the operative_config.gin file that I copied from the pretrained model directory, which I've been able to fine-tune on)".

Are you referring to a different model configuration via gin than the operative_config.gin file?

**Comment by craffel** [comment]
The operative config for the pre-trained model contains a line that specifies a pre-trained model checkpoint to load, which you don't want. You want appropriate gin files that set the configuration without using the operative config. The training command I linked to specifies such gin files.

**Comment by chrisrytting** [comment]
Sorry to circle back to this after such a long time, but I followed your advice and tried to point to a model configuration via gin. I tried following the example you linked to in the README and parsed the gin file corresponding to the model I'd like to use (T5-3B, but unpretrained) as opposed to the `operative_config.gin` of the pretrained model. To do this, I use the following code before instantiating a model and calling `train` on it:
```
gin.add_config_file_search_path('T5/t5/models/gin')
with gin.unlock_config():
    gin.parse_config_file('models/t5.1.0.3B.gin')
```
and got the error
```
Traceback (most recent call last):
  File "t5-train-from-scratch.py", line 274, in <module>
    steps=FINETUNE_STEPS,
  File "/usr/local/lib/python3.6/dist-packages/t5/models/mtf_model.py", line 229
, in train
    self.estimator(vocabulary, init_checkpoint),
  File "/usr/local/lib/python3.6/dist-packages/t5/models/mtf_model.py", line 207
, in estimator
    score_in_predict_mode=score_in_predict_mode)
  File "/usr/local/lib/python3.6/dist-packages/mesh_tensorflow/transformer/utils
.py", line 1710, in get_estimator
    mesh_shape=mesh_shape)
  File "/usr/local/lib/python3.6/dist-packages/mesh_tensorflow/transformer/utils
.py", line 210, in build_model
    layout=layout_rules)
  File "/usr/local/lib/python3.6/dist-packages/gin/config.py", line 1069, in gin
_wrapper
    utils.augment_exception_message_and_reraise(e, err_str)
  File "/usr/local/lib/python3.6/dist-packages/gin/utils.py", line 41, in augmen
t_exception_message_and_reraise
    raise proxy.with_traceback(exception.__traceback__) from None
  File "/usr/local/lib/python3.6/dist-packages/gin/config.py", line 1046, in gin
_wrapper
    return fn(*new_args, **new_kwargs)
  File "/usr/local/lib/python3.6/dist-packages/mesh_tensorflow/transformer/trans
former.py", line 1962, in make_bitransformer
    layer_stack=make_layer_stack(),
  File "/usr/local/lib/python3.6/dist-packages/gin/config.py", line 1039, in gin
_wrapper
    raise RuntimeError(err_str)
RuntimeError: Required bindings for `make_layer_stack` not provided in config: [
'layers']
  In call to configurable 'make_bitransformer' (<function make_bitransformer at
0x7f6096c3a7b8>)
```
This felt similar to #30, so I followed your instructions from there and, since I'm running everything locally and not on colab, downloaded the `operative_config.gin` for T5-3B from the pretrained models bucket for T5-3B on GCP and placed it in the directory I want to store my model in and parsed it there with 
```
gin.add_config_file_search_path('models-t5/3Bnpt')
with gin.unlock_config():
    gin.parse_config_file('operative_config.gin')
```
which gave me a different error
```
2021-05-19 21:57:27.415503: I tensorflow/stream_executor/platform/default/dso_lo
ader.cc:49] Successfully opened dynamic library libcudart.so.11.0
ERROR:root:Path not found: operative_config.gin
Traceback (most recent call last):
  File "t5-train-from-scratch.py", line 21, in <module>
    gin.parse_config_file('operative_config.gin')
  File "/usr/local/lib/python3.6/dist-packages/gin/config.py", line 1764, in par
se_config_file
    includes, imports = parse_config(f, skip_unknown=skip_unknown)
  File "/usr/local/lib/python3.6/dist-packages/gin/config.py", line 1678, in par
se_config
    __import__(statement.module)
  File "/usr/lib/python3.6/contextlib.py", line 99, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/local/lib/python3.6/dist-packages/gin/utils.py", line 58, in try_wi
th_location
    augment_exception_message_and_reraise(exception, _format_location(location))
  File "/usr/local/lib/python3.6/dist-packages/gin/utils.py", line 41, in augmen
t_exception_message_and_reraise
    raise proxy.with_traceback(exception.__traceback__) from None
  File "/usr/local/lib/python3.6/dist-packages/gin/utils.py", line 56, in try_wi
th_location
    yield
  File "/usr/local/lib/python3.6/dist-packages/gin/config.py", line 1678, in par
se_config
    __import__(statement.module)
ModuleNotFoundError: No module named 't5.data.sentencepiece_vocabulary'
  In file "models-t5/3Bnpt/operative_config.gin", line 2
    import t5.data.sentencepiece_vocabulary
```
This second instance feels like a bug, since `t5.data.sentencepiece_vocabulary` is deprecated according to `t5/models/mesh_transformer.py`. (And I know it says `ERROR:root:Path not found: operative_config.gin`, but it's definitely finding and reading that file)

Am I using the wrong gin configs, or parsing them in the wrong way, or is this a bug?

**Comment by craffel** [comment]
@adarob @sharannarang The operative configs on GCS still do
`import t5.data.sentencepiece_vocabulary`
Does that need to be changed?

**Comment by chrisrytting** [comment]
Check this out 

https://github.com/google-research/text-to-text-transfer-transformer/blob/abe601a171569aa0c8da790605f04d50db9a2f0a/t5/models/mesh_transformer.py#L28-L39

Seems like the idea is to skip importing these packages, but it apparently doesn't in my T5 (version 0.9.1, installed via pip). 

Can you tell me whether one of the above approaches sounds like it _should_ work? Seems like the `operative_config.gin` from the pretrained models has the right configuration as opposed to `models/t5.1.0.3B.gin`, which only has a few lines.

**Comment by chrisrytting** [comment]
I figured out my problem. After including `skip_unkown=t5.models.mesh_transformer.DEPRECATED_GIN_REFERENCES` in my call to `gin.parse_config_file` above, like 

```
import gin

gin.add_config_file_search_path('models-t5/3Bnpt')
with gin.unlock_config():
    gin.parse_config_file('operative_config.gin', skip_unknown=t5.models.mesh_transformer.DEPRECATED_GIN_REFERENCES)
```

I successfully started training a model from scratch! So thanks for your help!

Side note: I just copied the following code

https://github.com/google-research/text-to-text-transfer-transformer/blob/abe601a171569aa0c8da790605f04d50db9a2f0a/t5/models/mtf_model.py#L36-L40

so this should have worked after I specified the model directory upon instantiating the `MtfModel` object and having the `operative_config.gin` file in that directory, right? Cause this was the first thing I tried and it didn't work, but the only difference here is that I'm manually parsing the `operative_config.gin` file in my own script instead of letting it happen in the T5 code. Just wanted to let you guys know about that behavior.

I'm also confused as to why using the `models/t5.1.0.3B.gin` config doesn't work. Does that mean I couldn't have trained a T5-3B model from scratch on my own task without having access to a pretrained model?

Thanks for all you do!
